src/createEventEmitterProxy.ts

```diff
@@ -0,0 +1,136 @@
+import type { SwappableProxy } from './types';
+
+/**
+ * A portion of Node's EventEmitter interface that `createEventEmitterProxy`
+ * expects its `target` to support.
+ */
+type EventEmitterLike = {
+  eventNames: () => (string | symbol)[];
+  // The `rawListeners` method returns an array of `Function`s.
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  rawListeners(eventName: string | symbol): Function[];
+  removeAllListeners(event?: string | symbol): EventEmitterLike;
+};
+
+const filterNoop = () => true;
+const internalEvents: (string | symbol)[] = ['newListener', 'removeListener'];
+const externalEventFilter = (name: string | symbol) =>
+  !internalEvents.includes(name);
+
+/**
+ * Creates a proxy object that initially points to the given object but whose
+ * target can be substituted with another object later using its `setTarget`
+ * method. In addition, when the target is changed, event listeners which have
+ * been attached to the target will be detached and migrated to the new target.
+ *
+ * @template T - An object that implements at least `eventNames`, `rawListeners`,
+ * and `removeAllListeners` from [Node's EventEmitter interface](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/events.d.ts).
+ * @param initialTarget - The initial object you want to wrap.
+ * @param opts - The options.
+ * @param opts.eventFilter - Usually, listeners for all events will be migrated
+ * from old targets to new targets, but this function can be used to select
+ * which events you want to migrate. If you pass `'skipInternal'`, then
+ * `newListener` and `removeListener` will be excluded.
+ * @returns The proxy object.
+ */
+export function createEventEmitterProxy<T extends EventEmitterLike>(
```

Comment:
I know it's common to name generics like this, but there's no reason to not name generics with a more descriptive name. In this case, either would be fine, but once we start introducing more generics, it can get confusing pretty quickly.

```suggestion
export function createEventEmitterProxy<Type extends EventEmitterLike>(
```

Reply:
Yup! I didn't know if this was something we were starting to consistently do now, but it seems reasonable.